price/pricing/network_load_calculator.go

```diff
@@ -0,0 +1,242 @@
+	return &NetworkLoadMultiplierCalculator{
+		promClient:        promClient,
+		db:                db,
+		stop:              make(chan struct{}),
+		cachedMultipliers: make(map[ISO3166CountryCode]float64),
+	}
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) GetMultiplier(isoCode ISO3166CountryCode) float64 {
+	fetched, ok := nlmc.getCachedMulitplier(isoCode)
+	if !ok {
+		log.Error().Msgf("could not fetch multiplier for country %s", isoCode)
+		return 1
+	}
+	return fetched
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) setMultiplierMapCache(new map[ISO3166CountryCode]float64) {
+	nlmc.lock.Lock()
+	defer nlmc.lock.Unlock()
+
+	nlmc.cachedMultipliers = new
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) getCachedMulitplier(isoCode ISO3166CountryCode) (float64, bool) {
+	nlmc.lock.Lock()
+	defer nlmc.lock.Unlock()
+
+	v, ok := nlmc.cachedMultipliers[isoCode]
+	return v, ok
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) Start() error {
+	log.Info().Msgf("NetworkLoadMultiplierCalculator loading initial country map")
+	multis, err := nlmc.fetchMultipliersFromDB()
+	if err != nil {
+		return err
+	}
+
+	nlmc.setMultiplierMapCache(multis)
+	log.Info().Msgf("NetworkLoadMultiplierCalculator initial map loaded")
+
+	go func() {
+		log.Info().Msgf("NetworkLoadMultiplierCalculator started")
+		for {
+			select {
+			case <-nlmc.stop:
+				log.Info().Msgf("NetworkLoadMultiplierCalculator stopped")
+				return
+			case <-time.After(time.Minute * 30):
+				nlmc.updateCountries()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) Stop() {
+	nlmc.stopOne.Do(func() { close(nlmc.stop) })
+}
+
+type multiplierDB struct {
+	Multiplier  float64 `json:"multiplier"`
+	CountryCode string  `json:"country_code"`
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) fetchMultipliersFromDB() (map[ISO3166CountryCode]float64, error) {
+	conn, err := nlmc.db.Connection()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Release()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	rows, err := conn.Query(ctx, "SELECT multiplier, country_code FROM country_price_multipliers")
+	if err != nil {
+		return nil, err
+	}
+	results := make(map[ISO3166CountryCode]float64)
+	for rows.Next() {
+		var res = multiplierDB{}
+		if err := pgxscan.ScanRow(&res, rows); err != nil {
+			return nil, err
+		}
+		code := ISO3166CountryCode(res.CountryCode)
+		if code.Validate() != nil {
+			log.Error().Msgf("invalid country code %s in db", code)
+			continue
+		}
+		results[code] = res.Multiplier
+	}
+
+	return results, nil
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) updateCountries() {
+	log.Debug().Msgf("country update started")
+	newCache := make(map[ISO3166CountryCode]float64)
+	for k := range CountryCodeToName {
+		mul, err := nlmc.updateMultiplierForCountryByLoad(k)
+		if err != nil {
+			log.Err(err).Msgf("could not update multiplier for %s", k)
```

Comment:
This feels like it should actually not only log but outright return from here. Right now we'll have multiplier as 0, which is probably wrong, if we continue that country will be skipped which is not great either?